codegen/gce/src/expressions.rs

```diff
@@ -0,0 +1,364 @@
+use super::error::ConstraintEvaluationError;
+use super::{
+    utils::{
+        get_constant_index_by_matrix_access, get_constant_index_by_name,
+        get_constant_index_by_value, get_constant_index_by_vector_access, get_public_input_index,
+        get_random_value_index,
+    },
+    ExpressionJson, ExpressionOperation, NodeReference, NodeType,
+};
+use ir::{
+    constraints::{ConstantValue, Operation},
+    AirIR, NodeIndex,
+};
+use std::collections::BTreeMap;
+
+const MAIN_TRACE_SEGMENT_INDEX: u8 = 0;
+
+pub struct ExpressionsHandler<'a> {
+    ir: &'a AirIR,
+    constants: &'a [u64],
+    // maps indexes in Node vector in AlgebraicGraph and in `expressions` JSON array
+    expressions_map: &'a mut BTreeMap<usize, usize>,
+}
+
+impl<'a> ExpressionsHandler<'a> {
+    pub fn new(
+        ir: &'a AirIR,
+        constants: &'a [u64],
+        expressions_map: &'a mut BTreeMap<usize, usize>,
+    ) -> Self {
+        ExpressionsHandler {
+            ir,
+            constants,
+            expressions_map,
+        }
+    }
+
+    /// Parses expressions in transition graph's Node vector, creates [Expression] instances and pushes
+    /// them to the `expressions` vector.
+    pub fn get_expressions(&mut self) -> Result<Vec<ExpressionJson>, ConstraintEvaluationError> {
```

Comment:
We can change the signature to the following to use the builder pattern I suggested:

```
fn build_expressions(
        &mut self,
        ir: &AirIR,
        constants: &[u64],
    ) -> Result<(), ConstraintEvaluationError> {
```